url preview: Create a single preview for each URL in a message.

punchagan opened the pull request:
Fixes #8379.

Rebased #8419 against the latest master with a minor change - `processed_urls` is a `set` instead of a `list`.

timabbott commented on zerver/tests/fixtures/markdown_test_cases.json:
```diff
@@ -312,7 +312,7 @@
       "name": "two_inline_images",
       "input": "Google logo today: https://www.google.com/images/srpr/logo4w.png\nKinda boringGoogle logo today: https://www.google.com/images/srpr/logo4w.png\nKinda boring",
-      "expected_output": "<p>Google logo today: <a href=\"https://www.google.com/images/srpr/logo4w.png\" target=\"_blank\" title=\"https://www.google.com/images/srpr/logo4w.png\">https://www.google.com/images/srpr/logo4w.png</a><br>\nKinda boringGoogle logo today: <a href=\"https://www.google.com/images/srpr/logo4w.png\" target=\"_blank\" title=\"https://www.google.com/images/srpr/logo4w.png\">h ...
+      "expected_output": "<p>Google logo today: <a href=\"https://www.google.com/images/srpr/logo4w.png\" target=\"_blank\" title=\"https://www.google.com/images/srpr/logo4w.png\">https://www.google.com/images/srpr/logo4w.png</a><br>\nKinda boringGoogle logo today: <a href=\"https://www.google.com/images/srpr/logo4w.png\" target=\"_blank\" title=\"https://www.google.com/images/srpr/logo4w.png\">h ...
```
I think this test was intended to test two different images; maybe we should create a new test for that so we test both cases?

timabbott commented:
Awesome, thanks @punchagan!  Posted one comment, otherwise this lgtm.

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1011,10 +1011,17 @@ def run(self, root: Element) -> None:
         if len(found_urls) == 0 or len(found_urls) > self.INLINE_PREVIEW_LIMIT_PER_MESSAGE:
```
Oh, actually, one question -- should we be doing this number of URLs check after deduplication?

punchagan replied:
Yeah, I think that makes sense! I'll update the PR and let you know.

punchagan commented:
@timabbott I've updated the PR to change the test, and deduplicate URLs before checking if the message has more URLs than allowed.  

I've added a note to the commit message on the changes I made to the original PR, similar to how you add notes. Let me know if it's not the expected way for other contributors to do this.

punchagan commented:
@timabbott Pinging you on this PR, in case it was lost in your queue.

punchagan commented:
@timabbott I've updated this PR with another commit that disables previews for quoted links. This should address the concern raised in the comment [here](https://github.com/zulip/zulip/issues/11773#issuecomment-500493834). 

It seems like a reasonable thing to me to turn off previews for all quoted links, even if they weren't actually replies to messages with links.

timabbott commented:
I merged the first commit.  For the second, I'm pretty sure checking `.grandparent` is going to mean that a link in a bulleted list in a blockquote doesn't have the right behavior.  We probably need to walk the parent chain instead.

timabbott commented:
(And sorry for the slow review; this did indeed get a bit lost).

punchagan commented:
> For the second, I'm pretty sure checking `.grandparent` is going to mean that a link in a bulleted list in a blockquote doesn't have the right behavior. We probably need to walk the parent chain instead.

Ah, checking `.grandparent` didn't look right to me, but I didn't think too hard to generate counter examples. Thanks for the catch!

I've updated the PR to walk the parent chain to figure out if the link is inside a blockquote.

punchagan commented:
> (And sorry for the slow review; this did indeed get a bit lost).

No worries!

timabbott commented:
Looks great, merged thanks @punchagan!
